orca-retrofit/src/main/groovy/com/netflix/spinnaker/orca/retrofit/exceptions/BaseRetrofitExceptionHandler.groovy

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2023 Salesforce, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.retrofit.exceptions
+
+import java.lang.annotation.Annotation
+import java.lang.reflect.Method
+import com.netflix.spinnaker.orca.exceptions.ExceptionHandler
+import retrofit.http.RestMethod
+import static java.net.HttpURLConnection.*
+
+abstract class BaseRetrofitExceptionHandler implements ExceptionHandler {
+  boolean shouldRetry(Exception e, String kind, Integer responseCode) {
+    if (isMalformedRequest(kind, e.getMessage())) {
+      return false
+    }
+
+    // retry on 503 even for non-idempotent requests
+    if ("HTTP".equals(kind) && responseCode == HTTP_UNAVAILABLE) {
+      return true
+    }
+
+    return isIdempotentRequest(e) && (isNetworkError(kind) || isGatewayErrorCode(kind, responseCode) || isThrottle(kind, responseCode))
+  }
+
+  private boolean isGatewayErrorCode(String kind, Integer responseCode) {
+    "HTTP".equals(kind) && responseCode in [HTTP_BAD_GATEWAY, HTTP_UNAVAILABLE, HTTP_GATEWAY_TIMEOUT]
+  }
+
+  private static final int HTTP_TOO_MANY_REQUESTS = 429
+
+  boolean isThrottle(String kind, Integer responseCode) {
+    "HTTP".equals(kind) && responseCode == HTTP_TOO_MANY_REQUESTS
+  }
+
+  private boolean isNetworkError(kind) {
+    "NETWORK".equals(kind)
+  }
+
+  private boolean isMalformedRequest(String kind, String exceptionMessage) {
+    // We never want to retry errors like "Path parameter "blah" value must not be null.
+    return "UNEXPECTED".equals(kind) && exceptionMessage?.contains("Path parameter")
+  }
+
+  private static boolean isIdempotentRequest(Exception e) {
+    findHttpMethodAnnotation(e) in ["GET", "HEAD", "DELETE", "PUT"]
+  }
+
+  private static String findHttpMethodAnnotation(Exception exception) {
```

Comment:
Only worry, just at a glance here... there's a fair bit of reflection on this that I've seen cause major perf issues.  This particular code doesn't look bad off hand ... just one of those triggering a "caution" flag.  I'd wonder about using AnnotationUtils or similar spring libs which as i recall do caching of reflection calls so the forName/method search is only done once.

Reply:
That said... it IS a "move" soo.... ignore for now, just one of those caught my eye

Reply:
Yup, it's a drag, but it's no more reflection than we were already doing.  I don't think we need it once we move to retrofit2.